Make Style/RedundantSelf aware of arguments of a block

Envek opened the pull request:
Currently, Rubocop complains about redundant self when using a block with argument name same with receiver method, like this:

```rb
  %w[draft preview moderation approved rejected].each do |state|
    define_method "#{state}?" do
      self.state == state
    end
  end
```

Autocorrection fixes this code to `state == state` effectively breaking it.

This PR removes offense for arguments of a block with names equal to receiver method name.

-----------------

Before submitting the PR make sure the following are checked:

* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists). **Not exists**
* [x] Used the same coding conventions as the rest of the project.
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Added an entry to the [Changelog](../blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](../blob/master/CONTRIBUTING.md#changelog-entry-format).
* [x] All tests(`rake spec`) are passing.
* [x] The new code doesn't generate RuboCop offenses that are checked by `rake internal_investigation`.
* [x] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences.
* [x] Updated cop documentation with `rake generate_cops_documentation` (required only when you've added a new cop or changed the configuration/documentation of an existing cop). **Not required**

[1]: http://chris.beams.io/posts/git-commit/

Drenmi commented:
Nice catch, and nice PR. 🙂 Could you please add another example to the documentation and run `rake generate_cops_documentation`?

Envek commented:
I found that it doesn't handle nested blocks well (as in the first comment), so I changed implementation (and test) to deal with it.

Added example to documentation.

bbatsov commented:
I'm curious how is this different than passing a param from a method? Aren't block params lvars just as well?

bbatsov commented:
In particular I see that we already had this in the code:

```
def on_blockarg(node)
    on_argument(node)
end
```

Envek commented:
The main problem is that previous `add_scope` method cleared local variables for each nested scope that is fine for methods, because they can't be nested, but blocks usually are nested deeply.

Rebased on top of current master.
